File: 07-Banking-Kata/java/src/test/java/banking/service/CustomerServiceTest.java
package banking.service;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import banking.email.EmailService;
import banking.model.User;
import banking.persistence.NetworkRepository;
import banking.persistence.Repository;
import org.junit.Before;
import org.junit.Ignore;
import org.junit.Test;

import java.util.Optional;
import java.util.UUID;

public class CustomerServiceTest {

  private Repository<User> userRepository;
  private CustomerService customerService;
  private User user;
  private UUID userId;
  private MockEmailService mockEmailService;

  @Before
  public void setup() {
    user = new User("email@example.com");
    userId = user.getId();
    userRepository = new NetworkRepository<>();
    userRepository.save(user);

    mockEmailService = new MockEmailService();
    customerService = new CustomerService(userRepository, mockEmailService);
  }

  class MockEmailService implements EmailService {

    @Override
    public UUID sendMessage(String emailAddress, String message) {
      // Whatever fake implementation you choose...
      return null;
    }
  }

  @Test
  public void updateEmailSavesTheNewAddress() throws Exception {
    String expectedEmailAddress = "my-new-email@example.com";

    customerService.updateEmailAddress(userId, expectedEmailAddress);
    Optional<User> updatedUser = userRepository.findOne(userId);

    assertThat(updatedUser.get().getEmailAddress()).isEqualTo(expectedEmailAddress);
  }

  @Test
  public void updateEmailThrowsAnErrorForInvalidUserId() throws Exception {
    UUID invalidUserId = UUID.randomUUID();

    assertThatThrownBy(() -> {
      customerService.updateEmailAddress(invalidUserId, "my-new-email@example.com");
    }).hasMessage("Invalid userId");
  }

  @Test
  public void updateEmailSavesACopyOfTheUserWithoutModifyingTheOriginal() throws Exception {
    String expectedEmailAddress = "my-new-email@example.com";

    customerService.updateEmailAddress(userId, expectedEmailAddress);

    assertThat(user.getEmailAddress()).isNotEqualTo("my-new-email@example.com");
  }

  @Test
  @Ignore("Implement this test!")
  public void sendFraudAlertShouldNotifyTheUserByEmail() throws Exception {
    String expectedEmailAddress = "email@example.com";
    String expectedMessage = "You may have been the victim of fraud!";
    UUID expectedConfirmationCode = UUID.randomUUID();

    // TODO: use your mock here...

    UUID confirmationCode = customerService.sendFraudAlert(userId);

    assertThat(confirmationCode).isEqualTo(expectedConfirmationCode);
  }

  @Test
  @Ignore("Implement this test!")
  public void sendFraudAlertWillThrowAnErrorIfUserIdIsNotFound() throws Exception {
    // TODO
  }

}

File: 07-Banking-Kata/csharp-dotnetcore2/BankingKataAPI.Test/Services/CustomerServiceTest.cs
using System;
using System.Collections.Generic;
using System.Text;
using BankingKataAPI.Models;
using BankingKataAPI.Persistence;
using BankingKataAPI.Services;
using Microsoft.CodeAnalysis;
using Xunit;

namespace BankingKataAPI.Test.Services
{
    public class CustomerServiceTest
    {
        private readonly IRepository<User> _repository;
        private readonly CustomerService _service;
        private readonly User _user;
        private readonly Guid _userId;
        private readonly MockEmailService _emailService;

        public CustomerServiceTest()
        {
            _user = new User("example@email.com");
            _userId = _user.Id;

            _repository = new NetworkRepository<User>();
            _repository.Save(_user);

            _emailService = new MockEmailService();
            _service = new CustomerService(_repository, _emailService);
        }

        [Fact]
        public void UpdateEmailSavesTheNewAddress()
        {
            var expectedEmailAddress = "my-new-email@example.com";

            _service.UpdateEmailAddress(_userId, expectedEmailAddress);
            var updatedUser = _repository.FindOne(_userId);

            Assert.Equal(expectedEmailAddress, updatedUser.EmailAddress);
        }

        [Fact]
        public void UpdateEmailThrowsAnErrorForInvalidUserId()
        {
            var invalidUserId = Guid.NewGuid();

            var exception = Assert.Throws<Exception>(() => _service.UpdateEmailAddress(invalidUserId, "something@somewhere.com"));
            Assert.Equal("Invalid userId", exception.Message);
        }

        [Fact]
        public void UpdateEmailSavesACopyOfTheUserWithoutModifyingTheOriginal()
        {
            var expectedEmailAddress = "my-new-email@example.com";

            _service.UpdateEmailAddress(_userId, expectedEmailAddress);
            Assert.NotEqual(expectedEmailAddress, _user.EmailAddress);
        }

        [Fact(Skip = "Implement this feature")]
        public void SendFraudAlertShouldNotifyTheUserByEmail()
        {
            var expectedEmailAddress = "email@example.com";
            var expectedMessage = "You may have been the victim of fraud!";
            var expectedConfirmationCode = Guid.NewGuid();

            // TODO: use your mock here...

            var confirmationCode = _service.SendFraudAlert(_userId);

            Assert.Equal(expectedConfirmationCode, confirmationCode);
        }

        [Fact(Skip = "Implement this test")]
        public void SendFraudAlertWillThrowAnErrorIfUserIdIsNotFound()
        {

        }
    }

    public class MockEmailService : IEmailService
    {
        public Guid SendMessage(string emailAddress, string message)
        {
            return Guid.Empty;
        }
    }
}
